File: scripts/koruh/blocks/expunloader.js
const orblib = this.global.unity.exporb;

const expunloader = extendContent(Block, "exp-unloader", {
    load(){
        this.super$load();
        this.topRegion = Core.atlas.find(this.name + "-top");
        this.topRegion2 = Core.atlas.find(this.name + "-top2");
        this.sideRegion = [];
        for(var i=0; i<4; i++){
            this.sideRegion.push(Core.atlas.find(this.name + "-" + i));
        }
    },
    setStats(){
        this.super$setStats();
        this.stats.add(Stat.output, "@ [lightgray]@[]", Core.bundle.format("explib.expAmount", expunloader.unloadAmount * 10 * (60 / expunloader.unloadTime)), StatUnit.perSecond.localized());
    },
    noOrbCollision(){
        return true;
    }
});
expunloader.update = true;
expunloader.solid = true;
expunloader.timers = 1;
expunloader.unloadAmount = 2;
expunloader.unloadTime = 60;
expunloader.buildType = () => extend(Building, {
    _join: [false, false, false, false],
    draw(){
        this.super$draw();
        for(var i=0; i<4; i++){
            if(this._join[i]) Draw.rect(expunloader.sideRegion[i], this.x, this.y);
        }
        if(!this.consValid()) return;
        Draw.blend(Blending.additive);
        Draw.color(Color.white);
        Draw.alpha(Mathf.absin(Time.time(), 20, 0.4));
        Draw.rect(expunloader.topRegion, this.x, this.y);
        for(var i=0; i<4; i++){
            if(this._join[i]) Draw.rect(expunloader.topRegion2, this.x, this.y, i*90);
        }
        Draw.blend();
        Draw.reset();
    },
    updateTile(){
        this.super$updateTile();
        if(this.consValid() && this.enabled && this.timer.get(0, 60)){
            for(var i=0; i<4; i++){
                if(this._join[i]) this.checkUnload(i);
            }
        }
    },
    checkUnload(dir){
        var build = this.nearby(dir);
        if(build == null || !build.isValid() || !build.incExp){
            this._join[dir] = false;
            return;
        }
        for(var j=0; j<2; j++){
            if(build.totalExp() >= orblib.expAmount){
                build.incExp(-1 * orblib.expAmount);
                orblib.spewExp(this.x, this.y, 1, dir * 90 + 180, 8);
            }
            else break;
        }
    },
    onProximityUpdate(){
        this.super$onProximityUpdate();
        for(var i=0; i<4; i++){
            var build = this.nearby(i);
            this._join[i] = (build != null && build.isValid() && build.incExp && build.interactable(this.team));
        }
    }
});

// orblib.spewExp(this.x, this.y, 100, 6);

File: main/src/unity/world/blocks/storage/ExpUnloader.java
package unity.world.blocks.storage;

import arc.graphics.*;
import arc.graphics.g2d.*;
import arc.math.Mathf;
import arc.util.Time;
import mindustry.gen.Building;
import mindustry.world.Block;
import mindustry.world.meta.*;
import unity.entities.bullet.ExpOrb;
import unity.world.blocks.ExpBuildBase;

import static arc.Core.*;

public class ExpUnloader extends Block implements ExpOrbHandlerBase{
    protected float unloadAmount = 2f, unloadTime = 60f;
    protected TextureRegion topRegion, topRegion2;
    protected final TextureRegion[] sideRegions = new TextureRegion[4];

    public ExpUnloader(String name){
        super(name);
        update = solid = true;
        timers++;
    }

    @Override
    public void load(){
        super.load();
        topRegion = atlas.find(name + "-top");
        topRegion2 = atlas.find(name + "-top2");
        for(int i = 0; i < 4; i++) sideRegions[i] = atlas.find(name + "-" + i);
    }

    @Override
    public void setStats(){
        super.setStats();
        stats.add(Stat.output, "@ [lightgray]@[]", bundle.format("explib.expAmount", unloadAmount * 10f * (60f / unloadTime)), StatUnit.perSecond.localized());
    }

    public class ExpUnloaderBuild extends Building{
        protected final boolean[] join = new boolean[]{false, false, false, false};

        @Override
        public void draw(){
            super.draw();
            for(int i = 0; i < 4; i++){
                if(join[i]) Draw.rect(sideRegions[i], x, y);
            }
            if(!consValid()) return;
            Draw.blend(Blending.additive);
            Draw.color(Color.white);
            Draw.alpha(Mathf.absin(Time.time(), 20f, 0.4f));
            Draw.rect(topRegion, x, y);
            for(int i = 0; i < 4; i++){
                if(join[i]) Draw.rect(topRegion2, x, y, i * 90f);
            }
            Draw.blend();
            Draw.reset();
        }

        @Override
        public void updateTile(){
            if(enabled && consValid() && timer(0, 60f)){
                for(int i = 0; i < 4; i++){
                    if(join[i]) checkUnload(i);
                }
            }
        }

        protected void checkUnload(int dir){
            Building build = nearby(dir);
            if(!(build instanceof ExpBuildBase) || !build.isValid()){
                join[dir] = false;
                return;
            }
            ExpBuildBase temp = (ExpBuildBase)build;
            for(int i = 0; i < 2; i++){
                if(temp.totalExp() >= ExpOrb.expAmount){
                    temp.incExp(-1f * ExpOrb.expAmount);
                    ExpOrb.spewExp(x, y, 1, dir * 90f + 180f);
                }else break;
            }
        }

        @Override
        public void onProximityUpdate(){
            super.onProximityUpdate();
            for(int i = 0; i < 4; i++){
                Building build = nearby(i);
                join[i] = build instanceof ExpBuildBase && build.isValid() && build.interactable(team);
            }
        }
    }
}
